packages/plugins/src/phone/service.ts

```diff
@@ -0,0 +1,671 @@
+
+    if (checkExists) {
+      return err({
+        code: 'USER_ALREADY_EXISTS' as const,
+        message: 'Email or Phone already exists',
+      })
+    }
+
+    const count = await this.store.countActiveSignUpVerifications(data.phone)
+
+    // TODO: Customizable
+    // Check if the given phone number has reached the maximum OTP send limit
+    if (count >= 5) {
+      return err({
+        code: 'MAX_OTP_SEND_LIMIT_REACHED' as const,
+        message: 'Maximum OTP send limit reached. Please try again later.',
+      })
+    }
+
+    // Send phone verification OTP
+    const otpResponse = await this.options.signUp
+      .onOtpSent({
+        phone: data.phone,
+        name: data.name,
+        refCode: await this.store.generateRefCode(),
+      })
+      .then((result) => ok(result))
+      .catch((error) => err(error))
+
+    if (otpResponse.isErr()) {
+      return err({
+        code: 'FAILED_TO_SEND_OTP' as const,
+        message: 'Failed to send OTP',
+        cause: otpResponse.error,
+      })
+    }
+
+    const hashedPassword = await this.options.login.hashPassword(data.password)
+
+    const { password: _, ...rest } = data
+
+    const result = await this.store
+      .createSignUpVerification({
+        id: otpResponse.value.token,
+        refCode: otpResponse.value.refCode,
+        value: {
+          phone: data.phone,
+          password: hashedPassword,
+          data: rest as z.output<TSignUpBodySchema>,
+          attempt: 0,
+        },
+        // TODO: Check the expiration time from config
+        expiredAt: new Date(Date.now() + 5 * 60 * 1000), // 5 minutes
+      })
+      .then((result) => ok(result))
+      .catch((error) => err(error))
+
+    if (result.isErr()) {
+      return err({
+        code: 'FAILED_TO_CREATE_VERIFICATION' as const,
+        message: 'Failed to create sign up verification',
+        cause: result.error,
+      })
+    }
+
+    return ok({
+      token: otpResponse.value.token,
+      refCode: otpResponse.value.refCode,
+      resendAttempt: count + 1,
+      submitAttempt: 0,
+    })
+  }
+
+  async verifySignUpPhoneOtp(data: { phone: string; token: string; refCode: string; pin: string }) {
+    const verification = await this.store.getSignUpVerification(data.phone, data.refCode)
+
+    if (!verification) {
+      return err({
+        code: 'INVALID_OR_EXPIRED_VERIFICATION_TOKEN' as const,
+        message: 'Invalid or expired verification token',
+      })
+    }
+
+    // TODO: Customizable
+    if (verification.value.attempt >= 5) {
+      return err({
+        code: 'REACHED_MAX_ATTEMPTS' as const,
+        message: 'Maximum OTP verification attempts reached',
+      })
+    }
+
+    const verifyStatus = await this.options.signUp
+      .onOtpVerify({
+        phone: data.phone,
+        token: verification.id,
+        pin: data.pin,
+      })
+      .then((result) => ok(result))
+      .catch((error) => err(error))
```

Comment:
Can we use `fromPromise` instead ?

Reply:
Maybe like this
```ts
const result = await fromPromise(
  this.store.createSignUpVerification({
    id: otpResponse.value.token,
    refCode: otpResponse.value.refCode,
    value: {
      phone: data.phone,
      password: hashedPassword,
      data: rest as z.output<TSignUpBodySchema>,
      attempt: 0,
    },
    // TODO: Check the expiration time from config
    expiredAt: new Date(Date.now() + 5 * 60 * 1000), // 5 minutes
  }),
  R.identity()
)
```

Reply:
Sure, but the result is the same. I personally don't like `fromPromise` because of the indents